nntrainer/layers/rnncell.cpp

```diff
@@ -0,0 +1,238 @@
+    std::get<props::WeightInitializer>(*layer_impl_props);
+  Tensor::Initializer &bias_initializer = std::get<props::BiasInitializer>(*layer_impl_props);
+
+  unsigned int unit = std::get<props::Unit>(rnncell_props).get();
+  auto &hidden_state_activation_type =
+    std::get<props::HiddenStateActivation>(rnncell_props);
+  float dropout_rate = std::get<props::DropOutRate>(rnncell_props);
+  unsigned int max_timestep = std::get<props::MaxTimestep>(rnncell_props);
+
+  if (context.getNumInputs() != 1) {
+    throw std::invalid_argument("RNNCell layer takes only one input");
+  }
+
+  // input_dim = [ batch, 1, 1, feature_size ]
+  const TensorDim &input_dim = context.getInputDimensions()[0];
+  if (input_dim.channel() != 1 && input_dim.height() != 1) {
+    throw std::invalid_argument(
+      "Input must be single time dimension for RNNCell");
+  }
+  unsigned int batch_size = input_dim.batch();
+  unsigned int feature_size = input_dim.width();
+
+  // outut_dim = [ batch, 1, 1, hidden_size ( unit ) ]
+  TensorDim output_dim(batch_size, 1, 1, unit);
+
+  if (dropout_rate > epsilon) {
+    wt_idx[RNNCellParams::dropout_mask] = context.requestTensor(
+      output_dim, context.getName() + ":dropout_mask",
+      Tensor::Initializer::NONE, false, TensorLifespan::ITERATION_LIFESPAN);
+  }
+
+  context.setOutputDimensions({output_dim});
+
+  // weight_xh_dim : [1, 1, input_size, unit]
+  TensorDim weight_xh_dim({feature_size, unit});
+  // weight_hh_dim : [1, 1, unit, unit]
+  TensorDim weight_hh_dim({unit, unit});
+  // bias_h_dim : [1, 1, 1, unit]
+  TensorDim bias_h_dim({unit});
+
+  // weight_initializer can be set sepeartely. weight_xh initializer,
+  // weight_hh initializer kernel initializer & recurrent_initializer in keras
+  // for now, it is set same way.
+  wt_idx[RNNCellParams::weight_xh] = context.requestWeight(
+    weight_xh_dim, weight_initializer, weight_regularizer, weight_regularizer_constant,
+    context.getName() + ":weight_xh", true);
+  wt_idx[RNNCellParams::weight_hh] = context.requestWeight(
+    weight_hh_dim, weight_initializer, weight_regularizer, weight_regularizer_constant,
+    context.getName() + ":weight_hh", true);
+  wt_idx[RNNCellParams::bias_h] =
+    context.requestWeight(bias_h_dim, bias_initializer, WeightRegularizer::NONE,
+                          1.0f, context.getName() + ":bias_h", true);
+
+  // We do not need this if we reuse net_hidden[0]. But if we do, then the unit
+  // test will fail. Becuase it modifies the data during gradient calculation
+  // TODO : We could control with something like #define test to save memory
+  TensorDim dim(batch_size * max_timestep, 1, 1, unit);
+  wt_idx[RNNCellParams::hidden_state] = context.requestTensor(
+    dim, context.getName() + ":hidden_state", Tensor::Initializer::NONE, true,
+    TensorLifespan::ITERATION_LIFESPAN);
+
+  if (hidden_state_activation_type.get() == ActivationType::ACT_NONE) {
+    hidden_state_activation_type.set(ActivationType::ACT_TANH);
+  }
```

Comment:
looks like it's better that hidden_state_activation_type default is set to ACT_TANH instead

Reply:
The default hidden state activation type of RNN, LSTM, GRU which we've implemented is TANH but I'm not sure for other layers.

Reply:
Yeah, how about at the constructor we initialize with the default value?